core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java

```diff
@@ -327,6 +328,63 @@ public synchronized void losePrimacy() {
     LOG.info("Raft server successfully restarted");
   }
 
+  @Override
+  public synchronized SnapshotPResponse snapshot() {
+    SnapshotPResponse.Builder builder = SnapshotPResponse.newBuilder();
+    mSnapshotAllowed.set(true);
+    try {
+      CopycatClient client = createClient();
+      long sequence = ThreadLocalRandom.current().nextLong(Long.MIN_VALUE, 0);
+      CompletableFuture<Void> future = client.submit(new JournalEntryCommand(
+          JournalEntry.newBuilder().setSequenceNumber(sequence).build()));
+      try {
+        future.get(5, TimeUnit.SECONDS);
+      } catch (TimeoutException | ExecutionException e) {
+        LOG.info("Exception submitting entry: {}", e.toString());
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        return builder.setTriggered(false)
+            .setMessage("Interrupted while submitting entry to trigger snapshot").build();
+      }
+      if (!mStateMachine.isSnapshotting()) {
```

Comment:
could the snapshot have already finished?

Reply:
I tested on the alluxio cluster with fsmeta operations, the snapshot always fall into this if block. In this way, we are not exactly sure if `statemachine` snapshots. I will come up with other ways to make sure snapshot happens. You are right, the snapshot could have already finished.

Reply:
Add a `mLastSnapshotStartTime` to trigger when the last snapshot start. If it is bigger than the `RaftJournalSystem.checkpoint()` start time, we successfully trigger the snapshot.